Lib/profiling/sampling/stack_collector.py

```diff
@@ -287,3 +396,374 @@ def _create_flamegraph_html(self, data):
+                'max_samples': max_samples,
+                'percentage': 0  # Will be calculated after sorting
+            })
+
+        # Sort by total samples descending
+        file_stats.sort(key=lambda x: x['total_samples'], reverse=True)
+
+        # Calculate percentages relative to the max (top file = 100%)
+        if file_stats:
+            max_total = file_stats[0]['total_samples']
+            for stat in file_stats:
+                stat['percentage'] = (stat['total_samples'] / max_total * 100) if max_total > 0 else 0
+
+        # Build file index mapping filename -> html filename
+        self.file_index = {}
+        for i, stat in enumerate(file_stats):
+            safe_name = f"file_{i:04d}.html"
+            self.file_index[stat['filename']] = safe_name
+
+        # Generate individual file HTML reports
+        file_links = []
+        for i, stat in enumerate(file_stats):
+            safe_name = self.file_index[stat['filename']]
+            file_path = os.path.join(output_dir, safe_name)
+
+            self._generate_file_html(
+                file_path,
+                stat['filename'],
+                self.file_samples[stat['filename']],
+                stat
+            )
+
+            file_links.append({
+                'html_file': safe_name,
+                'stat': stat
+            })
+
+        # Generate index.html
+        index_path = os.path.join(output_dir, 'index.html')
+        self._generate_index_html(index_path, file_links)
+
+        print(f"Heatmap output written to {output_dir}/")
+        print(f"  - Index: {index_path}")
+        print(f"  - {len(file_links)} source file(s) analyzed")
+
+    def _generate_index_html(self, index_path, file_links):
+        """Generate index.html with list of all profiled files."""
+        import html
+        import base64
+        import importlib.resources
+
+        # Load template, CSS, JS and assets
+        template_dir = importlib.resources.files(__package__)
+        assets_dir = template_dir / "_assets"
+        template_content = (template_dir / "heatmap_index_template.html").read_text(encoding="utf-8")
+        css_content = (template_dir / "heatmap.css").read_text(encoding="utf-8")
+        js_content = (template_dir / "heatmap_index.js").read_text(encoding="utf-8")
+
+        # Load Python logo
+        png_path = assets_dir / "python-logo-only.png"
+        b64_logo = base64.b64encode(png_path.read_bytes()).decode("ascii")
+        logo_html = f'<img src="data:image/png;base64,{b64_logo}" alt="Python logo" class="python-logo"/>'
+
+        # Build table rows
+        table_rows = []
+        for link in file_links:
+            stat = link['stat']
+            full_path = html.escape(stat['filename'])
+            module_name = html.escape(stat['module_name'])
+            module_type = stat['module_type']
+            badge_class = f"badge-{module_type}"
+            bar_width = min(100, stat['percentage'] * 2)  # Scale for visibility
```

Comment:
I think this calculation isn't quite right? It looks like a file with 46.15% intensity is actually like 92% full because of the `* 2`


<img width="435" height="708" alt="Image" src="https://github.com/user-attachments/assets/25a66ee9-74cc-420d-afaf-a3a8734ade0f" />